### DISCUSSION TITLE
Is there a way to control the order of rendering?

### DISCUSSION THREAD

**Discussion by Dzejkop** [bevyengine/bevy#3773] [Q&A]
I have a mesh with some material, that the user will be able to paint on. I take the same mesh and add to it a semi transparent grid texture.

Thing is if I add some random component to this second mesh, then it's order of rendering changes. Before the first object painted over the grid. Now it paints over the object.

Is there any way to control it?

**Comment by cart** [comment]
I have a feeling that your entities are at the same z-level. Within a z-level, order is "undefined" and dependent on a number of factors. If you want a stable ordering between entities, just set `Transform::translation::z` to different values.

**Reply by MatrixDev** [reply]
@cart what if render order is needed for 3D? should I re-calculate translation and put it just a little closer to camera each time camera position changes?

**Reply by cart** [reply]
The StandardMaterial currently has a "depth bias" which can shift things closer or farther away from the camera. That would probably be easier than manually changing transforms. But that is only "local relative to current depth", which won't always be ideal. For this, we've discussed adding a "draw order" to the sort key that is "global".

**Reply by antilankin** [reply]
If you need to specify the rendering order independently of the object position you could use two cameras, as described in https://github.com/bevyengine/bevy/blob/main/examples/3d/two_passes.rs.
I use a similar technique to draw a semi-transparent selection marker in my map editor.

**Reply by cart** [reply]
I should have been more clear: "can shift things closer or farther away from the camera *when determining draw order*". Their actual transforms are unaffected.
